### ISSUE TITLE
Flatfield feature is broken: pink cast or does not work at all

### ISSUE THREAD

**Issue by AlexMatiash** [Beep6581/RawTherapee#3361]
Setup:
- image files Sony a7r II converted to DNG using latest Adobe DNG Converter
- the same for flatfield files OR non-compressed ARWs from a7r II too (obviously)
- RT 4.2.1074 x64 Release @ Win10 x64

In case of DNG flatfield files FF feature works, but adds pink cast in areas where it has effect. In case of ARW flatfield files FF feature does not work at all.

**Comment by heckflosse** [comment]
http://rawpedia.rawtherapee.com/How_to_write_useful_bug_reports

**Comment by AlexMatiash** [comment]
@heckflosse Well, and what have I missed from that list? I looked through the bugtracker and found nothing like this.

**Comment by Floessie** [comment]
@AlexMatiash Quoting from that site:

> Always make your raw file and PP3 file available to us by uploading them to a file sharing service such as www.filebin.net if your bug involves a crash on a particular raw file, a particular setting, or lack of support for your raw file.

By providing us with your problematic files, we might be able to reproduce and even fix your problem. ;)

**Comment by heckflosse** [comment]
@Floessie @AlexMatiash Maybe I should remove
`if your bug involves a crash on a particular raw file, a particular setting, or lack of support for your raw file.`

**Comment by AlexMatiash** [comment]
Sorry, my fault.

https://filebin.net/sa28ea5am53gnhw7  (LZMA2)

photo raw - DSC03959.dng, flatfield files - DSC06364.dng\arw.
Flatfield file is not perfect, but currently I have only it.

**Comment by Beep6581** [comment]
I updated http://rawpedia.rawtherapee.com/How_to_write_useful_bug_reports#Requirements_of_a_good_bug_report

**Comment by Beep6581** [comment]
The flat-field applied unto itself works perfectly.
When applied to the image there is discoloration around the periphery. I'd say the flat-field was shot incorrectly.
https://i.imgur.com/FyjA04i.jpg

**Comment by heckflosse** [comment]
@AlexMatiash cb05456 fixes the bug that your arw flat-field does not not work on your dng file

**Comment by iliasg** [comment]
@heckflosse Ingo, indeed the https://github.com/Beep6581/RawTherapee/commit/cb05456746066b05fd9504ac3a3cb545a5b0debe removes the special treatment A7RII & RX1 II get from Dcraw9.27 (remove right columns which contain garbage) and then Camconst.json takes care for a correct raw frame for both DNG and ARW. 
But this is a workaround, .. a correct solution would be to be able to define the correct frame in camconst.json independently from Dcraw, i.e. define "raw_crop" straight upon the raw frame (8000X5320 here) and not upon Dcraw's crop as it is now.
For Sony and Nikon FF sensors we cannot define absolute raw_crop values in camconst.json because they have dual modes (FF and APS-C) so we have to use the relative mode i.e.
"raw_crop": [ 0, 0, -36, 0 ], 
which should have the result to remove the 36 rightmost columns from the full raw frame. For the moment removes xx columns from the (possibly) already cropped by Dcraw frame.

**Comment by iliasg** [comment]
Closed by accident

**Comment by AlexMatiash** [comment]
> When applied to the image there is discoloration around the periphery. I'd say the flat-field was shot incorrectly.

But the same flatfield dng worked correctly either on arw photo (I can not test it now, because RT refuses to use dng ff to arw), or on the previous RT release. Also arw + arw pair works perfectly (as far as it is possible on such ff file) on the latest (4.2.1074 x64 Release) build.

https://filebin.net/41s8ozo6yg9r67qs here is that photo file in arw.

**Comment by heckflosse** [comment]
@AlexMatiash I applied your arw flat-field to your arw file from above filebin link => no colour cast using latest version. As you are using gtk3 version of rt you possibly have to wait some days to get a gtk3 version where it works correctly

**Comment by AlexMatiash** [comment]
> I applied your arw flat-field to your arw file from above filebin link => no colour cast using latest version.

Yes. The same is on my pc & RT. The problem is with the combination dng flatfield + dng photo.

**Comment by heckflosse** [comment]
@AlexMatiash Here's a screenshot: https://filebin.net/s458mx85smwfclnq/arwflatfield.png

**Comment by heckflosse** [comment]
@AlexMatiash `Yes. The same is on my pc & RT. The problem is with the combination dng flatfield + dng photo.`

Ok, maybe I was a bit confused :) I'll take a look again at this combination

**Comment by heckflosse** [comment]
@AlexMatiash I confirm that arw flat-field on arw file and dng flat-field on arw file work fine, but neither arw nor dng flat-field work fine on dng file.

**Comment by AlexMatiash** [comment]
@heckflosse Thank you! The reason to use dng instead of arw is in another (probably) bug of RT - HDD becomes crazy if I try to apply settings copied from one file to 2 or more other **uncompressed raw** files, in case when these files do not have .pp3 files. In case of compressed raw (it does not matter if it is compressed arw, cr2 or dng) everything is fine.

**Comment by heckflosse** [comment]
@AlexMatiash What's the amount of RAM in your pc?

**Comment by AlexMatiash** [comment]
@heckflosse 16 GB, no swap, and at least 8-10 GB free when I work with RT.

**Comment by strogonoff** [comment]
I have to concur, I’m having no luck with flat fields in RawTherapee whatsoever.

I’m working with DNGs out of Sigma FP, and trying to solve cyan-green vignetting from Voigtländer Color-Skopar 21mm f/3.5, possibly exacerbated by a CPL filter. I have tried many methods of taking flats (blank wall, into blue sky through various plastic diffusers, a sheet on top of white iPad screen, etc.) and made sure to follow the guidelines (match aperture, low ISO, etc.) as well as violate them hoping it would produce better results. Invariably, a flat applied unto itself worked perfectly as expected, but when applied to a photo it introduced strong red/magenta cast.

First, **a possible workaround for anyone struggling with this issue**: try an obscure little piece of software called [CornerFix](https://sites.google.com/site/cornerfix/). It can take a DNG, a flat field DNG, and produce _another DNG_ (all raw, mosaiced) which you can then process in RawTherapee as usual. GPL-licensed. I only just discovered it, but so far CornerFix worked well enough with the very same DNG combo that produced unusable results in RT.

(Notes: I don’t think CornerFix works on Linux, but despite being super old it worked successfully on macOS Catalina. I’m slightly concerned that DNG produced by CornerFix is more than 20% smaller than the original and will investigate that, quality seems equivalent so it might be just a losslessly compressed DNG.)

Back to RT’s issue. Below are resized renders of the same shot⁰ using RawTherapee 5.8 macOS build:

1. RT profile¹ applied (no flat field).

    ![_SDI3786-3](https://user-images.githubusercontent.com/292959/113548009-cbaefb00-9629-11eb-8b40-88b072298051.jpg)

2. Same RT profile, but with flat field² set.

    ![_SDI3786-2](https://user-images.githubusercontent.com/292959/113548021-cea9eb80-9629-11eb-94c2-530786acd069.jpg)

3. Same RT profile (no flat field), applied to DNG³ produced by CornerFix using the same flat field file.

    ![_SDI3786_CF-4](https://user-images.githubusercontent.com/292959/113548097-eb462380-9629-11eb-99a1-79b7a5b3217c.jpg)

⁰ [Original](https://s3.ap-northeast-2.amazonaws.com/a14f.com/rtff/_SDI3786.DNG)
¹ [RT profile](https://s3.ap-northeast-2.amazonaws.com/a14f.com/rtff/_SDI3786_CF-4.jpg.out.pp3) and [DCP profile](https://s3.ap-northeast-2.amazonaws.com/a14f.com/rtff/profile-sigma-skopar-sun.dcp) referenced in it
² [Flat field](https://s3.ap-northeast-2.amazonaws.com/a14f.com/rtff/SigmaFP-ColorSkopar-f8.DNG)
³ [DNG produced by CornerFix](https://s3.ap-northeast-2.amazonaws.com/a14f.com/rtff/_SDI3786_CF.DNG)

**Comment by heckflosse** [comment]
Can not reproduce in current dev version

![grafik](https://user-images.githubusercontent.com/13733487/113557676-b1135c80-95fe-11eb-8bd9-d77d668bd667.png)

**Comment by strogonoff** [comment]
> Can not reproduce in current dev version
> 
> ![grafik](https://user-images.githubusercontent.com/13733487/113557676-b1135c80-95fe-11eb-8bd9-d77d668bd667.png)

Excellent, definitely a bug in 5.8 then.
